obvs/patchscope.py

```diff
@@ -186,18 +191,34 @@ def source_forward_pass(self) -> None:
                 # TODO: validate this with non GPT2 & GPTJ models
                 self.source_model.transformer.wte.output = self.source.soft_prompt
 
-            self._source_hidden_state = self.manipulate_source().save()
-            self.source_output = self.source_model.lm_head.output[0].save()
+            self._source_hidden_state = self.manipulate_source().detach().save()
+            self.source_output = self.source_model.lm_head.output[0].detach().save()
 
     def manipulate_source(self) -> torch.Tensor:
         """
         Get the hidden state from the source representation.
 
         NB: This is seperated out from the source_forward_pass method to allow for batching.
         """
-        return getattr(getattr(self.source_model, self.MODEL_SOURCE), self.LAYER_SOURCE)[
+
+        # get the specified layer
+        layer = getattr(getattr(self.source_model, self.MODEL_SOURCE), self.LAYER_SOURCE)[
             self.source.layer
-        ].output[0][:, self._source_position, :]
+        ]
+
+        # if a head index is given, need to access the ATTN and HEAD components
+        if self.source.head is not None:
+            attn = getattr(layer, self.ATTN_SOURCE)
+            # TODO may not be .input for other models
+            head_act = getattr(attn, self.HEAD_SOURCE).input[0][0]
```

Comment:
Why are we using `input` instead of `output`?

My understanding is that `patchscope` always uses `output`, and if a researcher needs an input from layer `i`, they can access the output from layer `i-1`.

Reply:
The problem is, that the output of the c_attn layer in GPT2Attention is not the same as the input of the c_proj.
c_attn.output gets us the Q,K & Values concatenated together into one tensor. We want the attention layer outputs (sometimes referred to as z-values), which are calculated inbetween the c_attn and c_proj forward calls in the GPT2Attention object. So they are input of c_proj, but not output of c_attn
See GPT2Attention.forward for reference (https://github.com/huggingface/transformers/blob/main/src/transformers/models/gpt2/modeling_gpt2.py#L306)

Reply:
I.. see. So, `c_proj` is the equivalence of `W_O` in the original transformer paper?

If so, I agree that the concatenated head would be at `.attn.c_proj.input`.

![image](https://github.com/obvslib/obvs/assets/3963702/7283c04b-8931-44de-a0cb-4e34ddb642fd)

https://arxiv.org/pdf/1706.03762.pdf